pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -0,0 +1,389 @@
+			cd := e.Object.(*hivev1.ClusterDeployment)
+			return cd.Spec.ClusterPoolRef != nil && !cd.Spec.ClusterPoolRef.Claimed
+		},
+	}
+
+	clusterPoolMapFn := handler.ToRequestsFunc(
+		func(a handler.MapObject) []reconcile.Request {
+			var requests []reconcile.Request
+
+			cd := a.Object.(*hivev1.ClusterDeployment)
+			if cd.Spec.ClusterPoolRef != nil {
+				nsName := types.NamespacedName{Namespace: cd.Spec.ClusterPoolRef.Namespace, Name: cd.Spec.ClusterPoolRef.Name}
+				requests = append(requests, reconcile.Request{
+					NamespacedName: nsName,
+				})
+			}
+			return requests
+		})
+	err = c.Watch(
+		&source.Kind{Type: &hivev1.ClusterDeployment{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: clusterPoolMapFn,
+		},
+		clusterPredicate)
+	if err != nil {
+		return err
+	}
+
+	r.(*ReconcileClusterPool).dynamicClient, err = dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	r.(*ReconcileClusterPool).discoveryClient, err = discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	// TODO: watch ClusterDeployment, map to it's owning Pool (if any) and queue it up
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileClusterPool{}
+
+// ReconcileClusterPool reconciles a CLusterLeasePool object
+type ReconcileClusterPool struct {
+	client.Client
+	scheme          *runtime.Scheme
+	dynamicClient   dynamic.Interface
+	discoveryClient discovery.DiscoveryInterface
+}
+
+// Reconcile reads that state ClusterPool, checks if we currently have enough ClusterDeployments waiting, and
+// attempts to reach the desired state if not.
+func (r *ReconcileClusterPool) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := log.WithFields(log.Fields{
+		"clusterPool": request.Name,
+		"controller":  ControllerName,
+	})
+
+	logger.Infof("reconciling cluster pool: %v", request.Name)
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(ControllerName).Observe(dur.Seconds())
+		logger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the ClusterPool instance
+	clp := &hivev1.ClusterPool{}
+	err := r.Get(context.TODO(), request.NamespacedName, clp)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("pool not found")
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.WithError(err).Error("error reading cluster pool")
+		return reconcile.Result{}, err
+	}
+
+	// If the pool is deleted, do not reconcile.
+	if clp.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// Find all ClusterDeployments from this pool:
+	allCDs := &hivev1.ClusterDeploymentList{}
+	if err := r.Client.List(context.Background(), allCDs); err != nil {
```

Comment:
Hmm I thought we were removing that in favor of the explicit ref, are we sure we want two methods for this? My thought was we've already got all the CDs cached in memory as well so we could live without the label selector.

Reply:
I think it will be helpful to have a label to be able to quickly see all of the ClusterDeployments for a ClusterPool. I don't expect ClusterPools to share names that frequently. The fields in the spec are for source of truth: The label is for user-experience. With that being said, I am fine with omitting the label, too.

Reply:
Ok deferring for now on the label.